### PR TITLE
Adds a sample of how Crossplane + OAM work with KPT

### PR DESCRIPTION
**Pull request by wonderflow**
This PR adds a sample of how OAM work with [KPT](https://github.com/GoogleContainerTools/kpt).

### CODE REVIEW

**Comment by resouer** [comment]
The wording need to be updated to more "open" and "inspirational". I left some suggestions.

File: 5.OAM_KPT_Demo/README.md [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+              name:  example-appconfig-trait
+            spec:
+              replicaCount: 3
+```
+
+With KPT setters, this feature become more clear and useful. And you don't need to declare parameters in OAM yamls. Let's go through with it.
+
+#### Create setter by App Developer
+
+As the app developer need to add two parameter for his component, we add two KPT setters here:
```
Comment (resouer) [inline 5.OAM_KPT_Demo/README.md:174]:
```suggestion
Let's say the developer need to claim two fields as "configurable" for his application, he can add two kpt setters here:
```

File: 5.OAM_KPT_Demo/README.md [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+  NAME              VALUE               SET BY                   DESCRIPTION             COUNT
+  image           nginx:1.16.1        sampleapp developer   use to set image for component   0
+  instance-name   example-component   sampleapp developer   use to set an instance name      1
+```
+
+It's very clear and easy to understand.
+
+#### Set Value by App Operator
+
+Then the App Operator could set `instance-name` with a new name like this:
```
Comment (resouer) [inline 5.OAM_KPT_Demo/README.md:197]:
```suggestion
Then the application operator could set `instance-name` with a new name like this:
```

File: 5.OAM_KPT_Demo/README.md [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+
+### Install sample app
+
+```
+$ kubectl apply -f sampleapp/
+applicationconfiguration.core.oam.dev/example-appconfig created
+component.core.oam.dev/example-component created
+traitdefinition.core.oam.dev/manualscalertraits.core.oam.dev created
+workloaddefinition.core.oam.dev/containerizedworkloads.core.oam.dev created
+```
```
Comment (resouer) [inline 5.OAM_KPT_Demo/README.md:75]: Should explain what are the above resources. e.g.

- `Component` is the description of your application.
- `ApplicationConfiguration` is the description of operational policies for your `Component`.
- `TraitDefinition` and `WorkloadDefinition` are internal CRDs for OAM plugin to use.

File: 5.OAM_KPT_Demo/README.md [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+resources failed to the reached Current status
+traitdefinition.core.oam.dev/manualscalertraits.core.oam.dev pruned
+configmap/grouping-object-a4e18a93 pruned
+2 resource(s) pruned
+```
+
+Currently `kpt live apply` seems not work well and we have created an [issue #498](https://github.com/GoogleContainerTools/kpt/issues/498).
+
+
+### Plugin and Pipeline
```
[the remark this answers is not in the document]
Comment (wonderflow) [reply 5.OAM_KPT_Demo/README.md:289]: OK

**Comment by wonderflow** [comment]
@resouer  all clear, thanks so much for your advice!

**Comment by wonderflow** [comment]
@resouer  can you help merge it?

**Comment by wonderflow** [comment]
I have also sent a PR to KPT community https://github.com/GoogleContainerTools/kpt/pull/521

File: 5.OAM_KPT_Demo/README.md [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+
+With the heart of Configuration-as-Data, a typical Kubernetes native application management workflow just looks like a "pipeline". See the picture below:
+
+![img](./resource/kpt_oam.png)
+
+In this workflow, kpt is the manipulator of data. Stored in data source like Git, the original data (e.g. deployment.yaml) will pass through a pipeline of kpt functionalities to be manipulated into the desire state step by step. For example, `labels` added, `replicas` modified and `image` updated etc. 
+
+So what is OAM then? OAM is the format of data. More accurately, OAM is a complementary data format to Kubernetes so it can expose higher level abstractions such as "application" to developers.
```
Comment (resouer) [inline 5.OAM_KPT_Demo/README.md:19]:
```suggestion
So what is OAM then? OAM is the format of data. More accurately, OAM is a specific data format so Kubernetes can expose higher level abstraction such as "application" to developers.
```

File: 5.OAM_KPT_Demo/README.md [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+
+### Parameter Setting
+
+[kpt setters](https://googlecontainertools.github.io/kpt/guides/consumer/set/) is a powerful feature which naturally matches to the idea of "separate concerns" design from OAM.
+
+In Open Application Model, developers can claim certain fields in the application YAML as "configurable", so in the following workflow, operators (or the platform) will be allowed to modify these fields.
+
+Now this workflow can be easily achieved with help of kpt.
+
+OAM parameter is very important feature as it allows App Developer tell their demand to App Operator.
```
Comment (resouer) [inline 5.OAM_KPT_Demo/README.md:123]: Delete L123-186, as suggested by https://github.com/oam-dev/samples/pull/31#discussion_r406561090, how OAM set parameters is off topic for this blog post.

File: 5.OAM_KPT_Demo/README.md [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+
+```shell
+$ kpt cfg count sampleapp
+ApplicationConfiguration: 1
+Component: 1
+TraitDefinition: 1
+WorkloadDefinition: 1
+```
+
+So in the sampleapp, we have one ApplicationConfiguration, Component, TraitDefinition and WorkloadDefinition.
```
Comment (resouer) [inline 5.OAM_KPT_Demo/README.md:248]: It's better to make XxxDefinitions as part of Crossplane installation, ref: https://github.com/oam-dev/crossplane-oam-sample/issues/1

File: 5.OAM_KPT_Demo/README.md [hunk 8]
```diff
@@ -0,0 +1,281 @@
+# Managing custom Kubernetes applications with kpt
+
+This guidance will demonstrate how to use kpt to manage custom Kubernetes applications (i.e. apps defined by CRDs instead of raw Kubernetes workloads).
+
+We will use [Open Application Model](https://github.com/oam-dev/spec)(OAM) to describe the app for standardization and portability consideration. But you are free to use any Custom Resource to describe your application as well.
```
Comment (resouer) [inline 5.OAM_KPT_Demo/README.md:5]:
```suggestion
We will use [Open Application Model](https://github.com/oam-dev/spec)(OAM) to describe the app for standardization and portability consideration. But you are free to use any Custom Resource to describe your application as well. [Crossplane](https://github.com/crossplane/crossplane) will be installed as the OAM runtime for Kubernetes.
```

File: 5.OAM_KPT_Demo/README.md [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+$ kubectl apply -f sampleapp/
+component.core.oam.dev/example-component created
+applicationconfiguration.core.oam.dev/example-appconfig created
+traitdefinition.core.oam.dev/manualscalertraits.core.oam.dev created
+workloaddefinition.core.oam.dev/containerizedworkloads.core.oam.dev created
+```
+
+Several resources are created:
+
+* Component is the description of your application.
```
Comment (resouer) [inline 5.OAM_KPT_Demo/README.md:89]:
```suggestion
* Component is the description of what you want to deploy.
```

File: 5.OAM_KPT_Demo/README.md [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+
+In this workflow, kpt is the manipulator of data. Stored in data source like Git, the original data (e.g. deployment.yaml) will pass through a pipeline of kpt functionalities to be manipulated into the desire state step by step. For example, `labels` added, `replicas` modified and `image` updated etc. 
+
+So what is OAM then? OAM is the format of data. More accurately, OAM is a complementary data format to Kubernetes so it can expose higher level abstractions such as "application" to developers.
+
+In that sense, it's a natural match for using kpt to manipulate the data which is formatted by OAM specification.
+
+## Pre-requisites
+
+Install OAM plugin to Kubernetes. In this demo, we will use Crossplane by following [this installation guides](https://github.com/oam-dev/crossplane-oam-sample#installation).
```
Comment (resouer) [inline 5.OAM_KPT_Demo/README.md:25]:
```suggestion
Install OAM Kubernetes runtime (Crossplane) by following [its installation guides](https://github.com/oam-dev/crossplane-oam-sample#installation).
```

**Comment by wonderflow** [comment]
@resouer  all comments are fixed, please merge this https://github.com/oam-dev/crossplane-oam-sample/pull/2 to make XxxDefinitions as part of Crossplane installation

File: 5.OAM_KPT_Demo/README.md [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+fetching package /5.OAM_KPT_Demo/repository/sampleapp from https://github.com/oam-dev/samples to sampleapp
+```
+
+```shell
+➜  kpt tree sampleapp
+sampleapp
+├── Kptfile
+├── appconfig.yaml
+├── component.yaml
+├── traitdef.yaml
```
Comment (resouer) [inline 5.OAM_KPT_Demo/README.md:71]: XxxDefinition should not be part of the app. Please fix this and same issues below.

Comment (wonderflow) [reply 5.OAM_KPT_Demo/README.md]: sorry, fixed

File: 5.OAM_KPT_Demo/README.md [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+
+### App Overview
+
+With kpt, you could see an overview of your App.
+
+```shell
+$ kpt cfg count sampleapp
+ApplicationConfiguration: 1
+Component: 1
+TraitDefinition: 1
```
Comment (resouer) [inline 5.OAM_KPT_Demo/README.md:174]: Definition?

**Comment by resouer** [comment]
LGTM, please update the kpt side Pull Request.

**Pull request merged**